apps/choo/bootstrap/kernel.hoon

```diff
@@ -285,161 +315,381 @@
+::    .nodes: the nodes of the dependency graph
+::    .bc: the build cache
+::
+::    returns a trap vase with the compiled hoon file and the updated build cache
+++  compile-target
+  |=  [pat=path =path-dag nodes=(map path node) bc=build-cache]
+  ^-  [(trap vase) build-cache]
+  ~&  >>  compiling-target+pat
+  =/  n=node
+    ~|  """
+        couldn't find node {<pat>} in path-dag.
+        nodes: {<~(key by nodes)>}
+        path-dag: {<~(key by path-dag)>}
+        """
+    +:(~(got by path-dag) pat)
+  =/  graph  (build-graph-view nodes)
+  =/  next=(map path node)  (update-next nodes graph)
+  =|  vaz=(trap vase)
+  |-
+  ?:  .=(~ next)
+    (compile-node n path-dag bc)
+  =^  vaz  bc
+    %+  roll  ~(tap by next)
+    |=  [[p=path n=node] [v=_vaz bc=_bc]]
+    =^  new-vaz  bc  (compile-node n path-dag bc)
+    ?:  =(vaz *_vaz)
+      [new-vaz bc]
+    [(slew v new-vaz) bc]
+  %=  $
+    next   (update-next nodes graph)
+    graph  (roll ~(tap by next) |=([[p=path *] g=_graph] (update-graph-view g p)))
+    bc     bc
+    vaz    vaz
+  ==
+::
+::  $compile-node: compile a single node
+::
+::    .n: the node to compile
+::    .path-dag: the path-dag of the dependency graph
+::    .bc: the build cache
+::
+::    looks up the node in the build cache and compiles it if it's not already
+::    cached.
+::
+::    returns a trap vase with the compiled hoon and the updated build cache
+++  compile-node
+  |=  [n=node =path-dag bc=build-cache]
+  ^-  [(trap vase) build-cache]
+  ~&  >  compiling-node+path.n
+  =/  [dep-hash=@ *]  (~(got by path-dag) path.n)
+  ?:  (~(has by bc) dep-hash)
+    ~&  >  build-cache-hit+path.n
+    [(~(got by bc) dep-hash) bc]
+  ~&  >  build-cache-miss+path.n
+  =/  vaz=(trap vase)  (build-node n path-dag bc)
+  [vaz (~(put by bc) dep-hash vaz)]
+::
+::  $build-node: build a single node and its dependencies
+::
+::    .n: the node to compile
+::    .path-dag: the path-dag of the dependency graph
+::    .bc: the build cache
+::
+::    returns a trap vase with the compiled hoon
+++  build-node
+  |=  [n=node =path-dag bc=build-cache]
+  ^-  (trap vase)
+  =;  dep-vaz=(trap vase)
+    (swet (slew honc dep-vaz) hoon.n)
+  %+  roll
+    deps.n
+  |=  [raut vaz=(trap vase)]
+  ~&  >  grabbing-dep+pax
+  =/  [dep-hash=@ dep-node=node]
+    ~|  "couldn't find dep hash for {<pax>}"
+    (~(got by path-dag) pax)
+  =/  dep-vaz=(trap vase)
+    %+  fall  (~(get by bc) dep-hash)
+    (build-node dep-node path-dag bc)
+  (slew (slew honc vaz) (label-vase dep-vaz face))
+::
+::  $label-vase: label a (trap vase) with a face
+::
+::    .vaz: the (trap vase) to label
+::    .face: the face to label the (trap vase) with
+::
+::    returns a (trap vase) labeled with the given face
+++  label-vase
+  |=  [vaz=(trap vase) face=(unit @tas)]
+  ^-  (trap vase)
+  ?~  face  vaz
+  =>  [vaz=vaz face=u.face]
+  |.
+  =/  vas  $:vaz
+  [[%face face p.vas] q.vas]
+::
+::
++$  merk-dag  (map @ node)
++$  path-dag  (map path [@ node])
```

Comment:
Can you put comments on these two types explaining how they're constructed and used?

Reply:
how's this?

Reply:
Great!